include/ttmlir/Dialect/TTNN/Transforms/Workarounds/Decomposition/ReduceOpsRewritePattern.h

```diff
@@ -0,0 +1,105 @@
+getReduceDims(const std::optional<mlir::ArrayAttr> &dimArg);
+
+// Calculates the shape of the new Reduce op created in the workaround, based
+// on the input shape and reducing dimensions.
+llvm::SmallVector<int64_t>
+calculateNewReduceShape(RankedTensorType inputType,
+                        const std::optional<mlir::ArrayAttr> &dimArg);
+
+// Creates the dimArg attribute of the new Reduce op created in the workaround.
+// In case when reduce is done over all dimensions of the input nullptr is
+// returned, because Metal supports reduce over all dimensions for any tensor
+// rank when reduce dimensions are not specified, but it doesn't support reduce
+// for tensors with rank larger than 2 when reduce dimensions are specified.
+mlir::ArrayAttr
+createNewReduceDimArg(RankedTensorType inputType,
+                      const std::optional<mlir::ArrayAttr> &dimArg);
+
+// This workaround addresses next two Metal issues:
+// - https://github.com/tenstorrent/tt-metal/issues/13361
+// - https://github.com/tenstorrent/tt-metal/issues/16118
+//
+// TODO(mrakita): Remove this workaround once these Metal issues are fixed
+// (tracked by https://github.com/tenstorrent/tt-mlir/issues/1624).
+//
+template <typename ReduceOp>
+class ReduceOpsRewritePattern : public OpRewritePattern<ReduceOp> {
+public:
+  using OpRewritePattern<ReduceOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(ReduceOp srcOp,
+                                PatternRewriter &rewriter) const override {
+    if (srcOp.getKeepDim()) {
+      return failure();
+    }
+
+    RankedTensorType inputType = srcOp.getInput().getType();
+    RankedTensorType outputType = srcOp.getResult().getType();
+
+    ReduceOp newReduceOp =
+        createReduceOpWithKeepDim(srcOp, rewriter, inputType, outputType);
+
+    // Metal TTNN implementation of Reduce ops doesn't yet support
+    // keepDim=false. As a workaround, we convert Reduce ops to combination of
+    // Reduce op with keepDim=true + Reshape op to remove the reduce dims so
+    // that the rest of the graph is not affected. In case when this is not
+    // needed (for example because type converters already promoted rank of the
+    // op result) then we avoid adding unnecessary Reshape op.
+    if (outputType.getShape().size() < inputType.getShape().size()) {
+      replaceOpWithReshapeOp(srcOp, newReduceOp, rewriter, outputType);
+    } else {
+      rewriter.replaceOp(srcOp, newReduceOp);
+    }
+
+    return success();
+  }
+
+private:
+  ReduceOp createReduceOpWithKeepDim(ReduceOp srcOp, PatternRewriter &rewriter,
+                                     RankedTensorType inputType,
+                                     RankedTensorType outputType) const {
+    llvm::SmallVector<int64_t> outputShapeVec =
+        calculateNewReduceShape(inputType, srcOp.getDimArg());
+
+    RankedTensorType newOutputType = RankedTensorType::get(
+        outputShapeVec, inputType.getElementType(), inputType.getEncoding());
```

Comment:
Just double check if layout encoding depends on the outputShapeVec?

Reply:
Yeah that is exactly the thing I wanted to discuss with you :) I will use here the utils you've mentioned yesterday to set the new shape to the output type, instead of using the input type encoding. But we should still discuss the implications of this, maybe some decompositions will require more of the layout logic re-done.

Reply:
You can take a look at the helper TTNNLayoutAttr functions that we have:
![image](https://github.com/user-attachments/assets/c5afb065-8d05-4d22-9d58-8d2a9dca9baa)

We don't have one that will recreate the layout attr with the new shape, but it should be easy to add a new one here as well...